File: gtci/00-warmup/cpp/3-is-palindrome.cpp
#include <cctype>
#include <iostream>

class Solution {
   public:
    bool isPalindrome(std::string s) {
        std::string::iterator left = s.begin();
        std::string::iterator right = s.end();

        while (left < right) {
            while (left < right && !isalnum(*left)) left++;

            while (left < right && !isalnum(*right)) right--;

            if (tolower(*left) != tolower(*right)) return false;

            left++;
            right--;
        }

        return true;
    }
};

#include <cassert>

int main() {
    Solution solution;
    assert(solution.isPalindrome("A man, a plan, a canal: Panama") == true);
    assert(solution.isPalindrome("Was it a car or a cat I saw?") == true);
    assert(solution.isPalindrome("race a car") == false);
    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/00-warmup/python3/3-is-palindrome.py
def isPalindrome(s: str) -> bool:
    left = 0
    right = len(s) - 1

    while left < right:
        while left < right and not s[left].isalnum():
            left += 1

        while left < right and not s[right].isalnum():
            right -= 1

        if s[left].lower() != s[right].lower():
            return False

        left += 1
        right -= 1

    return True


if __name__ == "__main__":
    testString = "A man, a plan, a canal: Panama"
    if isPalindrome(testString):
        print(f'"{testString}" is a palindrome.')
    else:
        print(f'"{testString}" is not a palindrome.')

    anotherTestString = "Was it a car or a cat I saw?"
    if isPalindrome(anotherTestString):
        print(f'"{anotherTestString}" is a palindrome.')
    else:
        print(f'"{anotherTestString}" is not a palindrome.')
